DIPs/DIP1NNN-DK.md

```diff
@@ -0,0 +1,617 @@
+- [Ownsership and borrowing in D](https://dlang.org/blog/2019/07/15/ownership-and-borrowing-in-d/)
+- [DIP1021](https://github.com/dlang/DIPs/blob/master/DIPs/accepted/DIP1021.md)
+- [my vision of D's future](https://dlang.org/blog/2019/10/15/my-vision-of-ds-future/)
+
+Take reference counting for example: As long as the reference count can be meddled with from `@safe` code, freeing the memory can't be `@trusted` and D won't support `@safe` reference counting.
+
+### Existing holes in `@safe`
+
+While constructing arbitrary pointers inside `@safe` code is not allowed, there is currently no defence around pointers initialized to an unsafe value outside any function:
+```D
+@safe:
+int* x = cast(int*) 0x7FFE_E800_0000;
+
+void main() {
+    *x = 3; // memory corruption
+}
+```
+
+Another issue is that, unlike pointers, the `bool` type is regarded `@safe` to overlap or void-initialize.
+The problem here is that the optimizer may assume it is always `0` or `1`, but the bool type can be any `ubyte` value in practice.
+By constructing a `bool` that is larger than 1 and indexing it in an array with compile-time length 2, memory can be corrupted since the range check is elided.
+See issue 19968: [@safe code can create invalid bools resulting in memory corruption](https://issues.dlang.org/show_bug.cgi?id=19968)
+
+The issue is marked fixed after DMD PR [#10055](https://github.com/dlang/dmd/pull/10055) was merged, which defensively inserts a bitwise 'and' operation (`b & 1`) when `bool` values are promoted to integers.
+This is not a complete solution however, since there are also other ways in which invalid booleans give undefined behavior: [void initializated bool can be both true and false](https://issues.dlang.org/show_bug.cgi?id=20148).
+This happens because negation of a `bool` is simply toggling the least significant bit with the xor operator (`b ^ 1`), which only works if all other bits are 0.
+This could, again, be fixed by inserting more `& 1` instructions, but it starts to defeat the purpose of a specialized `bool` primitive type when it can't be guaranteed to be `0` or `1`: you might as well make `bool` a wrapper around `ubyte` then.
+
+And even if you accept this solution for booleans, the solution of ensuring validity on usage can't work on pointers, since there is no simple operation that type checks a pointer at run-time.
+A proper solution should make it impossible for invalid values of types to even reach `@safe` code.
+
+## Prior work
+
+I brought up this limitation of `@trusted` code [on the newsgroup](https://forum.dlang.org/thread/lobjvmjxvvmamklzfzhp@forum.dlang.org) originally with a different solution in mind.
+User 'ag0aep6g' suggested:
+
+> My best idea so far (might not be originally mine) is to let the @system attribute apply to variables.
+
+Any prior discussion about `@system` variables has not been found.
```

Comment:
Some previous mentions of the idea:

* https://forum.dlang.org/post/cxupcgybvqwvkuvcokoz@forum.dlang.org (Olivier FAURE)
* https://forum.dlang.org/post/pqleml$2kpg$1@digitalmars.com (Timon Gehr)
* https://github.com/dlang/dmd/pull/8035#issuecomment-373921399 (myself)
* https://forum.dlang.org/post/pqn0dc$2cq7$1@digitalmars.com (myself)

That last one contains an alternative solution to the problem that isn't mentioned in this DIP: an `Unsafe` type wrapper that obscures the stored value, e.g. with `void[T.sizeof]`, so that all accesses become `@system`. It's probably not good for precise garbage collection, though.

Reply:
Thanks for the links, I'll add them.

Reply:
The `void[T.sizeof]` trick does not work right? You may not retrieve pointers with it, but you can still write to them in `@safe` code.
```D
@safe: 
void[int.sizeof] unsafeVar; // hides a pointer

union U {
    int asInt = 3;
    ubyte[int.sizeof] asBytes;
}

void main() {
    unsafeVar = U.init.asBytes;
}
```

Reply:
Hmm, from https://github.com/dlang/dmd/pull/8035#issuecomment-373921399 it seems like Walter wants to make going around `private` not `@safe` after all. I might need to spend more time considering that alternative.

Reply:
> The `void[T.sizeof]` trick does not work right? You may not retrieve pointers with it, but you can still write to them in `@safe` code.

Huh. I completely missed that. Looks like the `void[T.sizeof]` idea is a no-go.

Writing through a `void[]` that way also seems to be allowed, and that's obviously unsafe. Filed as <https://issues.dlang.org/show_bug.cgi?id=20345>. But `void[]`'s problem is that it might alias pointers. `void[n]` can't do that in `@safe` code (I think). So it's not exactly the same issue.